connector/src/main/scala/com/microsoft/kusto/spark/datasink/KustoWriter.scala

```diff
@@ -189,94 +158,164 @@ object KustoWriter{
-          case OperationStatus.Succeeded =>
-            KDSU.logInfo(myName, s"Ingestion to Kusto succeeded. Cluster: '$cluster', database: '$database', table: '$table'$batchIdIfExists, partition: '$partitionId'")
-          case otherStatus =>
-            throw new RuntimeException(s"Ingestion to Kusto failed with status '$otherStatus'." +
-              s" Cluster: '$cluster', database: '$database', table: '$table'$batchIdIfExists, partition: '$partitionId'. Ingestion info: '${readIngestionResult(res)}'")
-        }).await(timeoutMillis, TimeUnit.MILLISECONDS)
-    }
+        }
+      },
+      timeout)
+  }
 
-  def All(list: util.ArrayList[Boolean]) : Boolean = {
+  def all(list: util.ArrayList[Boolean]): Boolean = {
     val it = list.iterator
     var res = true
-    while (it.hasNext && res)
-    {
+    while (it.hasNext && res) {
       res = it.next()
     }
     res
   }
 
-  @throws[IOException]
-  private[kusto] def serializeRows(rows: Iterator[InternalRow], schema: StructType, bos: BlobOutputStream, timeZone: String): Unit = {
-    val gzip = new GZIPOutputStream(bos)
+  def createFileWriter(schema: StructType,
+                       timeZone: String,
+                       tableCoordinates: KustoCoordinates,
+                       tmpTableName: String,
+                       storageUri: String,
+                       container: CloudBlobContainer): FileWriteResource = {
+    val blobName = s"${tableCoordinates.database}_${tmpTableName}_${UUID.randomUUID.toString}_SparkStreamUpload.gz"
+    val blob: CloudBlockBlob = container.getBlockBlobReference(blobName)
+    val gzip: GZIPOutputStream = new GZIPOutputStream(blob.openOutputStream())
     val csvSerializer = new KustoCsvSerializationUtils(schema, timeZone)
 
     val writer = new OutputStreamWriter(gzip, StandardCharsets.UTF_8)
-    val buffer:BufferedWriter = new BufferedWriter(writer, GZIP_BUFFER_SIZE)
-    val csvWriter = new CsvWriter(buffer, new CsvWriterSettings)
+    val buffer: BufferedWriter = new BufferedWriter(writer, GZIP_BUFFER_SIZE)
+    val csvWriter: CsvWriter = new CsvWriter(buffer, new CsvWriterSettings)
+    datasink.FileWriteResource(buffer, gzip, csvWriter, blob)
+  }
 
-    for (row <- rows) {
-      csvWriter.writeRow(csvSerializer.convertRow(row))
+  @throws[IOException]
+  private[kusto] def serializeRows(rows: Iterator[InternalRow],
+                                   storageUri: String,
+                                   parameters: KustoWriteResource): Seq[CloudBlockBlob] = {
+    import parameters._
+    val container: CloudBlobContainer = new CloudBlobContainer(new URI(storageUri))
+    val fileWriter: FileWriteResource = createFileWriter(schema, writeOptions.timeZone, coordinates, tmpTableName, storageUri, container)
+
+    val (_, _, blobs) = rows.foldLeft[(Long, FileWriteResource, Seq[CloudBlockBlob])]((0, fileWriter, Seq())) { case ((size, fileWriter, blobsCreated), row) =>
```

Comment:
it seems like an overkill for the foldleft function - as both variables fileWriter, blobsCreated dont change most of the operation . 
i think it will be better without the 'case'
and without declaring fileWriter before (u can simply return  (_ , fileWriter, blobs)  )

Reply:
I think it is a Scala idiom to prefer immutable code. 
I also think it prefers it because of the way it works behind scenes, so using val is usually more efficient (there are some cases that it is not, but not in ours I think). 
[link1](https://www.quora.com/Scala-Are-there-any-performance-implications-to-using-vars-versus-vals)
[link2](https://en.wikipedia.org/wiki/Scala_(programming_language)#Immutability)
So yes fileWriter can be a var and then we can update it later... But I don't see the need. 
The foldLeft has to start from something and it accumulates the values. This is why I define the fileWriter before that. I don't want to construct it every iteration, just when I need to. I am also using the case so I can give the tuple fields names to make  it more readable. I don't think it is less efficient or something, there are no types specified so I don't think it does like an extra check or something in that case. But I am not sure and didn't find any evidence to support that, just an intuition then...

Reply:
ok, your call. 
anyway, I  wasnt talking about it being var . The first line can simply be : 
    val (_, fileWriter, blobs) = rows.foldLeft[(Long, FileWriteResource, Seq[CloudBlockBlob])]((0, createFileWriter(schema, writeOptions.timeZone, coordinates, tmpTableName, storageUri, container), Seq())) { case ((size, fileWriter, blobsCreated), row) => ....

Reply:
ahhh I need it to be declared first also for finalizing it later in case the partition size was less than a GB

Reply:
I think its also more readable when created first... in any case adding some comments for future generations can be helpful here